From 0daa30db44093bd424da1db7c8110da0b751d1d5 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Thu, 23 Aug 2018 19:03:53 +0300
Subject: [PATCH 0849/1200] fix: ahci: Support disk insertion after boot

Run AHCI port initialization procedure when the port command
buffer is not allocated, but link is up.
This patch allows plugging in AHCI devices after board boot.

Change-Id: I94121c5ceeb7334d073eb572222b0b7c73828ada
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/59440
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Igal Liberman <igall@marvell.com>
---
 drivers/ata/ahci.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/ata/ahci.c b/drivers/ata/ahci.c
index 6aeb587147..9b9ebd236c 100644
--- a/drivers/ata/ahci.c
+++ b/drivers/ata/ahci.c
@@ -678,6 +678,17 @@ static int ahci_device_data_io(struct ahci_uc_priv *uc_priv, u8 port, u8 *fis,
 		return -1;
 	}
 
+	/*
+	 * If the device was plugged after boot, the port is not initialized
+	 * Try to restart the port for supporting device hot plug-in
+	 */
+	if (pp->cmd_tbl == 0) {
+		if (ahci_port_start(uc_priv, port)) {
+			printf("Cannot restart port %d\n", port);
+			return -1;
+		}
+	}
+
 	memcpy((unsigned char *)pp->cmd_tbl, fis, fis_len);
 
 	sg_count = ahci_fill_sg(uc_priv, port, buf, buf_len);
-- 
2.22.0

